### ISSUE TITLE
 onfly_checksum d,7c3731a3105dda4647eecbd3b0056fdc differ from message 7c3731a3105dda4647eecbd3b0056fdc

### ISSUE THREAD

**Issue by petersilva** [MetPX/sarracenia#158]
This cropped up in relation to work done to deal with ET-CTS requests to change the format of the messages.

The message in the title (where the *same* checksum sparks a warning as being different) is strange and indicates a subtle error somewhere.  in git repo (master) as part of testing the new v03 format, I set up a configuration that downloads from text files from the datamart, converts them to v03 inlined format, and then posts them on the local broker.   Then I have something that subscribe to the local v03 feed.   Once in a while, it prints a message about checksums being different, but the only difference is the presence of the algorithm indicatorr at the beginning of the onfly one.
hmm...

to reproduce:
```

cat >~/.config/sarra/subscribe/dd_amis.conf <<EOT
broker amqps://dd.weather.gc.ca/
instances 5
subtopic bulletins.alphanumeric.#

inline
inline_encoding text

directory /tmp/amis
accept .*

post_topic_prefix v03.post
post_broker amqp://tfeed@localhost
post_exchange xpublic
EOT

```
then: 

```
broker amqp://tsource@localhost

topic_prefix v03.post
exchange xpublic
reportback False

subtopic #

directory /tmp/all
accept .*

```
mkdir /tmp/all mkdir/tmp/amis

in one shell:

sr_subscribe foreground dd_amis 

in a second shell:

```
sr_subscribe foreground dd_all

in this shell you will see:
   .
   .
   .
019-02-22 15:34:45,582 [INFO] file_log downloaded to: /tmp/all/SAIR31_OIII_222030___56077
2019-02-22 15:34:45,584 [INFO] file_log downloaded to: /tmp/all/SASP35_LEMM_222030___56079
2019-02-22 15:34:45,585 [INFO] file_log downloaded to: /tmp/all/SRCN40_KWAL_222033___56084
2019-02-22 15:34:45,586 [WARNING] onfly_checksum d,7c3731a3105dda4647eecbd3b0056fdc differ from message 7c3731a3105dda4647eecbd3b0056fdc
2019-02-22 15:34:45,586 [INFO] file_log downloaded to: /tmp/all/SAUS70_KWBC_222034___56080
2019-02-22 15:34:45,589 [INFO] file_log downloaded to: /tmp/all/SRCN40_KWAL_222033___56082
2019-02-22 15:34:45,590 [INFO] file_log downloaded to: /tmp/all/SAFR31_LFPW_222030___56075
20
   .
   .
   .

```

The reason for the strange (spurious?) onfly_checksum message is the mystery to be solved.

**Comment by benlapETS** [comment]
is the second config for dd_all ?

**Comment by benlapETS** [comment]
I couldnt reproduce but looking in the code I may found a fix tell me if it does the job or not ??

**Comment by benlapETS** [comment]
Here is my python env:
```
# Name                    Version                   Build  Channel
amqp                      2.4.1                    pypi_0    pypi
appdirs                   1.4.3                    pypi_0    pypi
argh                      0.26.2                   pypi_0    pypi
asn1crypto                0.24.0                   py37_0  
bcrypt                    3.1.6            py37h7b6447c_0  
ca-certificates           2019.1.23                     0  
certifi                   2018.11.29               py37_0  
cffi                      1.11.5           py37he75722e_1  
cryptography              2.5              py37h1ba5d50_0  
humanize                  0.5.1                    pypi_0    pypi
idna                      2.8                      py37_0  
libedit                   3.1.20181209         hc058e9b_0  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 8.2.0                hdf63c60_1  
libsodium                 1.0.16               h1bed415_0  
libstdcxx-ng              8.2.0                hdf63c60_1  
metpx-sarracenia          2.19.2b1                  dev_0    <develop>
ncurses                   6.1                  he6710b0_1  
netifaces                 0.10.9                   pypi_0    pypi
openssl                   1.1.1a               h7b6447c_0  
paramiko                  2.4.2                    py37_0  
pathtools                 0.1.2                    pypi_0    pypi
pip                       19.0.1                   py37_0  
psutil                    5.5.1                    pypi_0    pypi
pyasn1                    0.4.5                      py_0  
pycparser                 2.19                     py37_0  
pyftpdlib                 1.5.4                    pypi_0    pypi
pynacl                    1.3.0            py37h7b6447c_0  
python                    3.7.2                h0371630_0  
pyyaml                    3.13                     pypi_0    pypi
readline                  7.0                  h7b6447c_5  
setuptools                40.8.0                   py37_0  
six                       1.12.0                   py37_0  
sqlite                    3.26.0               h7b6447c_0  
tk                        8.6.8                hbc83047_0  
vine                      1.2.0                    pypi_0    pypi
watchdog                  0.9.0                    pypi_0    pypi
wheel                     0.32.3                   py37_0  
xz                        5.2.4                h14c3975_4  
zlib                      1.2.11               h7b6447c_3  
```

**Comment by petersilva** [comment]
I think this commit fixed it:  530c8a1a4a0f32ec0116c110fe976098fd741d9e
released in v2.19.03b1

**Comment by benlapETS** [comment]
I saw that problem back yesterday, not fixed yet, im working on it...

**Comment by benlapETS** [comment]
I left a commit that seems to be good but not yet fully tested

**Comment by petersilva** [comment]
good stuff! might want to pull...  I committed something to: add *randomize* directive to watch/f40 config, so that sr_watch now generates a variety of checksums instead of just 's' .   the flow_test fails now, in subscribe cp_f61
```

2019-04-02 23:46:07,608 [DEBUG] sftp_transport close
2019-04-02 23:46:07,608 [DEBUG] sr_sftp close
2019-04-02 23:46:07,608 [DEBUG] sr_sftp init
2019-04-02 23:46:07,608 [DEBUG] iotime 300
2019-04-02 23:46:07,610 [ERROR] Download failed sftp://peter@localhost//home/peter/sarra_devdocroot/sent_by_tsource2send/20190403/WXO-DD/observations/swob-ml/20190403/CXKI/2019-04-03-0255-CXKI-AUTO-minute-swob.xml
2019-04-02 23:46:07,611 [DEBUG] Exception details: 
Traceback (most recent call last):
  File "/home/peter/src/sarracenia/sarra/sr_util.py", line 492, in download
    msg.onfly_checksum = "{},{}".format(proto.get_sumalgo(), proto.checksum)
  File "/home/peter/src/sarracenia/sarra/sr_util.py", line 344, in get_sumalgo
    return self.sumalgo.registered_as()
AttributeError: 'NoneType' object has no attribute 'registered_as'
2019-04-02 23:46:07,611 [DEBUG] publish AMQP is used
20
```

same thing happens in 
`
sr_subscribe -debug foreground u_sftp_f60
`
I think it is the same bug for both.

**Comment by petersilva** [comment]
it might be my fault, it it could be tickling one of those cases you were worried about.

**Comment by benlapETS** [comment]
That was my fault in fact. I didnt think that sumalgo could be None at this point. And there is a check for every other except in this case. So I adapted the code and that will be more coherent with other protos...

**Comment by petersilva** [comment]
I merged the code, and it gradually dawned on me why checksum calculation occurs in a few places for different reasons:

case 1) in sr_post/compute_sumstr:  making an initial checksum for initial announcement of a file. 
sample decision, if you get z,d here, you should leave it, as the intent is for people who need to read the file as part of transfer anyways calculate the sum on the fly as it downloads, rather than costing an extra read to the poster.

vs. 
case 2) receiving a message to decide whether to download...
sr_message/content_should_not_be_downloaded(

sample decision: if the upstream/message checksum is z,d it should always download because we just don know. You can´t read the file here because it is on the remote machine before transfer (in the subscribe/sarra case.)


between 2 and 3) I think there is a case that is currently missed: uploading
case 2.5)
Question: if a sender has z,d ... I think it just sends it as z,d... I´m wondering if an onfly_checksum should be calculated as part of the send, and the final message posted after delivery ought perhaps to be d,onfly_checksum. perhaps also glorified with on_data filtering, which makes it close to 
identicaly to case 3... I´m wondering if this should be an issue?  I doin't think anyone uses it this way, so it causes no practical issue, but it the meaning of z, is that one should re-calculate the checksum as soon as convenient. We are only deferring for efficiency reasons to avoid an extra read.  Since the sender has to read the data, it makes sense that it does the replacement.


vs. 
case 3) after downloading (in the relatively infinite do_download routine), when you have the actual data, you need to determine the correct checksum for storage (in the attribute), and for posting to those later down the chain.  

sample decision: If the message checksum is z,d then the checksum to persist and forward should be d,onfly_checksum value. 

other consideration:
There is also the case of an on_data plugin being present, where the data written locally does not match the data received.  so you track the data received with onfly_checksum, and the data written locally as data_checksum.   In this case, the checksum to compare with upstream to identify a transmission error is the onfly_checksum, but the one to persist and forward is the data_checksum.

Not saying it cannot be improved, just that there are reasons why it is the way it is, and it takes some thought.  It's when people ask questions that I start to understand the code.  It's a good thing.

**Comment by petersilva** [comment]
I think the original problem is solved, I wonder if a new issue should be opened for the sender case.
There is probably room to have case 2.5 and 3 be identical.
It might also add on_data support to sending.

**Comment by petersilva** [comment]
I goofed, I didn't run the flow_test before merging.  I could not get it to run clean after merging to master. so master was busted for a while. I'm getting 14/21 to max 16/21 on repeated runs.  I decided to comment out *randomize* in the sarra/examples/watch/f40.conf and the next run all is fine... so... I think the patches you submitted are good, but the randomize is a sort a cheap code fuzzing, and it seems to have tickled other problems.
Do you want to keep working on that here? or do we open another issue for turning *randomize* back on in watch.

master is clean again.

**Comment by petersilva** [comment]
fix released v2.19.09 perhaps even included in v2.19.04b2
